subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/plugins/JavaGradlePluginPlugin.java

```diff
@@ -256,9 +255,9 @@ public void execute(Task taskWithAction) {
 
             task.getOutputFile().set(project.getLayout().getBuildDirectory().file("reports/plugin-development/validation-report.txt"));
 
-            task.getClasses().setFrom(mainSourceSet.getOutput().getClassesDirs());
-            task.getClasspath().setFrom(mainSourceSet.getCompileClasspath());
-            task.dependsOn(mainSourceSet.getOutput());
+            task.getClasses().setFrom((Callable<Object>) () -> extension.getPluginSourceSet().getOutput().getClassesDirs());
+            task.getClasspath().setFrom((Callable<Object>) () -> extension.getPluginSourceSet().getCompileClasspath());
+            task.dependsOn((Callable<Object>) () -> extension.getPluginSourceSet().getOutput());
```

Comment:
Isn't the dependency automatically added by the classes? Given that `classesDirs.builtBy(this)`? https://github.com/gradle/gradle/blob/8a082ed93daba60c41b4687b429ac9a3c0397464/subprojects/plugins/src/main/java/org/gradle/api/internal/tasks/DefaultSourceSetOutput.java#L49-L51

Reply:
Good point, I removed it.